src/NHibernate/Transaction/AdoNetWithSystemTransactionFactory.cs

```diff
@@ -0,0 +1,384 @@
+				_locked = true;
+				_semaphore = new SemaphoreSlim(0);
+			}
+
+			private void Unlock()
+			{
+				// Do not set _locked back to false, we want to wait only once per transaction context.
+				// Volatile member, do not simplify by removing the local variable.
+				var semaphore = _semaphore;
+				_semaphore = null;
+				if (semaphore == null)
+					return;
+				// Not supposed to have more than one waiting thread, but in case the user made bad multi-threading,
+				// do not leave frozen threads.
+				semaphore.Release(100);
+				semaphore.Dispose();
+			}
+
+			#region IEnlistmentNotification Members
+
+			void IEnlistmentNotification.Prepare(PreparingEnlistment preparingEnlistment)
+			{
+				using (new SessionIdLoggingContext(_sessionImplementor.SessionId))
+				{
+					try
+					{
+						using (_sessionImplementor.ConnectionManager.BeginsProcessingFromSystemTransaction(_useConnectionOnSystemTransactionEvents))
+						{
+							if (_useConnectionOnSystemTransactionEvents)
+							{
+								using (var tx = new TransactionScope(AmbientTransaction))
+								{
+									_sessionImplementor.BeforeTransactionCompletion(null);
+									foreach (var dependentSession in _sessionImplementor.ConnectionManager.DependentSessions)
+										dependentSession.BeforeTransactionCompletion(null);
+
+									tx.Complete();
+								}
+							}
+							else
+							{
+								_sessionImplementor.BeforeTransactionCompletion(null);
+								foreach (var dependentSession in _sessionImplementor.ConnectionManager.DependentSessions)
+									dependentSession.BeforeTransactionCompletion(null);
+							}
+						}
+						// Lock the session to ensure second phase gets done before the session is used by code following
+						// the transaction scope disposal.
+						Lock();
+
+						_logger.Debug("Prepared for system transaction");
+						preparingEnlistment.Prepared();
+					}
+					catch (Exception exception)
+					{
+						_logger.Error("System transaction prepare phase failed", exception);
+						preparingEnlistment.ForceRollback(exception);
+					}
+				}
+			}
+
+			void IEnlistmentNotification.Commit(Enlistment enlistment)
+				=> ProcessSecondPhase(enlistment, true);
+
+			void IEnlistmentNotification.Rollback(Enlistment enlistment)
+				=> ProcessSecondPhase(enlistment, false);
+
+			void IEnlistmentNotification.InDoubt(Enlistment enlistment)
+				=> ProcessSecondPhase(enlistment, null);
+
+			private void ProcessSecondPhase(Enlistment enlistment, bool? success)
+			{
+				using (new SessionIdLoggingContext(_sessionImplementor.SessionId))
+				{
+					_logger.Debug(
+						success.HasValue
+							? success.Value
+								? "Committing system transaction"
+								: "Rolled back system transaction"
+							: "System transaction is in doubt");
+					// we have not much to do here, since it is the actual
+					// DB connection that will commit/rollback the transaction
+					// After transaction actions are raised from TransactionCompleted event.
```

Comment:
No more "transaction completion" from "In-Doubt" case. This case trigger transaction completion too, without delay. That was probably by lack of understanding of that case that the code was previously calling transaction completion from "In-Doubt", just in case. But since the 2PC can be run concurrently to the transaction completion event, that was just messing the whole thing.